{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>\n",
       "code_show=true; \n",
       "function code_toggle() {\n",
       " if (code_show){\n",
       " $('div.input').hide();\n",
       " } else {\n",
       " $('div.input').show();\n",
       " }\n",
       " code_show = !code_show\n",
       "} \n",
       "$( document ).ready(code_toggle);\n",
       "</script>\n",
       "The raw code for this IPython notebook is by default hidden for easier reading.\n",
       "To toggle on/off the raw code, click <a href=\"javascript:code_toggle()\">here</a>."
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from py2neo import Graph\n",
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "import networkx as nx\n",
    "from IPython.display import IFrame, display, HTML\n",
    "import tempfile\n",
    "import ipywidgets as widgets\n",
    "from ipywidgets import interact, interact_manual\n",
    "from networkx.readwrite import json_graph\n",
    "from cyjupyter import Cytoscape\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "import logging\n",
    "logger = logging.getLogger()\n",
    "logger.setLevel(logging.CRITICAL)\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "The raw code for this IPython notebook is by default hidden for easier reading.\n",
    "To toggle on/off the raw code, click <a href=\"javascript:code_toggle()\">here</a>.''')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "driver = Graph(\"http://141.5.106.249:7474/db/data/\", user=\"neo4j\", password=\"bioinfo1112\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Zoo Proteome \n",
    "\n",
    "Proteins preform the vast majority of functions in all biological domains but their large-scale investigation has lagged behind for technological reasons. Since the first essentially complete eukaryotic proteome was reported1, advances in mass spectrometry (MS)-based proteomics2 have enabled increasingly comprehensive identification and quantification of the human proteome 3456. However, there are few comparisons across species, especially compared to genomics initiatives 7. Here, we employ an advanced proteomics workflow, in which the peptide separation step is performed by a microstructured and extremely reproducible chromatographic system, for the in-depth measurement of 100 taxonomically diverse organisms. With two million peptide and 340,000 stringent protein identifications obtained in a standardized manner, we double the number of proteins with solid experimental evidence known to the scientific community. The data also provide a foundation for machine learning, as we demonstrate by experimentally confirming predicted peptide properties of bacteroides uniformis. Our results provide a comparative view into the functional organization of organisms across the entire evolutionary range. A remarkably high fraction of the total proteome mass in all kingdoms is dedicated to protein homeostasis and folding, highlighting the challenge of maintaining protein structure across all of life. Likewise, a constantly high fraction is involved in supplying energy resources, although the pathways range from photosynthesis through iron sulphur metabolism to carbohydrate metabolism.!\n",
    "\n",
    "![alt text](Figure_1_JoMu.tif \"Zoo Proteome\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "254bfd41b45248139db09d830c5aa56f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='num_species', options=(25, 50, 100), value=25), Dropdown(descripti…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def visualize_nx(path):\n",
    "    plt.figure(figsize=(15,15)) \n",
    "    pos=nx.spring_layout(path)\n",
    "    nx.draw_networkx(path, pos,node_size=460,font_size=12)\n",
    "    plt.axis('off');\n",
    "    \n",
    "def visualize_cys(path):\n",
    "    pathcys, layout, stylesheet = networkx_to_cytoscape(path)\n",
    "    pathvis = Cytoscape(data={'elements':pathcys}, visual_style=stylesheet, layout=layout)\n",
    "\n",
    "    return pathvis\n",
    "\n",
    "def get_nx_graph(query_path):\n",
    "    G = None\n",
    "    regex = r\"\\(?(.+)\\)\\<?\\-\\>?\\[\\:(.+)\\s\\{.*\\}\\]\\<?\\-\\>?\\((.+)\\)?\"\n",
    "    nodes = set()\n",
    "    rels = set()\n",
    "    for r in query_path:\n",
    "        path = str(r['path'])\n",
    "        matches = re.search(regex, path)\n",
    "        if matches:\n",
    "            source = matches.group(1)\n",
    "            source_match = re.search(regex, source)\n",
    "            if source_match:\n",
    "                source = source_match.group(1)\n",
    "                relationship = source_match.group(2)\n",
    "                target = source_match.group(3)\n",
    "                nodes.update([source, target])\n",
    "                rels.add((source, target, relationship))\n",
    "                source = target\n",
    "            relationship = matches.group(2)\n",
    "            target = matches.group(3)\n",
    "            nodes.update([source, target])\n",
    "            rels.add((source, target, relationship))\n",
    "    if len(nodes)> 0:\n",
    "        G = nx.Graph()\n",
    "        G.add_nodes_from(nodes, color='blue')\n",
    "        for s,t,label in rels:\n",
    "            G.add_edge(s,t,label=label, width=1)\n",
    "\n",
    "    return G\n",
    "\n",
    "def networkx_to_cytoscape(graph):\n",
    "    cy_graph = json_graph.cytoscape_data(graph)\n",
    "    cy_nodes = cy_graph['elements']['nodes']\n",
    "    cy_edges = cy_graph['elements']['edges']\n",
    "    cy_elements = cy_nodes\n",
    "    cy_elements.extend(cy_edges)\n",
    "    layout = {'name': 'cose',\n",
    "                'idealEdgeLength': 100,\n",
    "                'nodeOverlap': '4',\n",
    "                'refresh': 20,\n",
    "                'fit': True,\n",
    "                #'padding': 30,\n",
    "                'randomize': True,\n",
    "                'componentSpacing': 100,\n",
    "                'nodeRepulsion': 400000,\n",
    "                'edgeElasticity': 100,\n",
    "                'nestingFactor': 5,\n",
    "                'gravity': '10',\n",
    "                'numIter': 1000,\n",
    "                'initialTemp': 200,\n",
    "                'coolingFactor': 0.95,\n",
    "                'minTemp': 1.0,\n",
    "                 'height':'1000','width':'1400'}\n",
    "\n",
    "    stylesheet = [{'selector': 'node', 'style': {'label': 'data(name)', 'width':5, 'height':5,'font-size':1, 'background-color':'#3182bd'}}, \n",
    "                {'selector':'edge','style':{'curve-style': 'bezier', 'width':0.5}}]\n",
    "\n",
    "    return cy_elements, layout, stylesheet\n",
    "\n",
    "\n",
    "    \n",
    "@interact\n",
    "def generate_graph(num_species=[25,50,100], visualize=['Cytoscape', 'Networkx']):\n",
    "    query = '''MATCH path=(p:Project)-[:STUDIES_SPECIE]-(s:Taxonomy{name:s.name})RETURN path LIMIT ''' + str(num_species)\n",
    "    query_path = driver.run(query).data()\n",
    "    \n",
    "    G = get_nx_graph(query_path)\n",
    "    \n",
    "    \n",
    "    if visualize == 'Cytoscape':\n",
    "        return visualize_cys(G)\n",
    "    else:\n",
    "        return visualize_nx(G)\n",
    "    \n",
    "\n",
    "    return None\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e11b9ea9fc3b49289d46de5732a273fe",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Text(value='O14983', description='query_protein'), Text(value='29760, 9615', description…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@interact\n",
    "def get_homologs(query_protein='O14983', query_species='29760, 9615', query_type=['All','Orthologs','Paralogs'], visualize=['Cytoscape', 'Networkx', 'Table']):\n",
    "    query= r'MATCH path=(p1:Protein)-[r:RELATIONSHIP]-(p2:Protein)-[:BELONGS_TO_TAXONOMY]-(t:Taxonomy) WHERE (p1.id=\"{}\" OR p1.name=\"{}\") AND p2.id<>p1.id'.format(query_protein, query_protein)\n",
    "    query_species = query_species.split(',')\n",
    "    query_protein_specie = r'MATCH (p:Protein)-[:BELONGS_TO_TAXONOMY]-(t:Taxonomy) WHERE (p.id=\"{}\" OR p.name=\"{}\") RETURN t.id AS taxid'.format(query_protein, query_protein)\n",
    "    \n",
    "    sp_request = driver.run(query_protein_specie).data()\n",
    "    spdf = pd.DataFrame(sp_request)\n",
    "    if not spdf.empty:\n",
    "        sp = spdf['taxid'][0]\n",
    "        query_species.append(sp)\n",
    "    if '' not in query_species or len(query_species)>1:\n",
    "        query = query + \" AND p2.taxid IN [{}]\".format(\",\".join(['\"{}\"'.format(s.strip()) for s in query_species]))\n",
    "    if query_type == \"Orthologs\":\n",
    "        query = query.replace(\"RELATIONSHIP\", 'IS_ORTHOLOG')\n",
    "    elif query_type == 'Paralogs':\n",
    "        query = query.replace(\"RELATIONSHIP\", 'IS_PARALOG')\n",
    "    else:\n",
    "        query = query.replace(\"RELATIONSHIP\", 'IS_ORTHOLOG|IS_PARALOG')\n",
    "    \n",
    "    if visualize == \"Networkx\" or visualize == \"Cytoscape\":\n",
    "        query = query + \"RETURN path\"\n",
    "        request = driver.run(query).data()\n",
    "        path = get_nx_graph(request)\n",
    "        if path is not None:\n",
    "            if visualize == \"Networkx\":\n",
    "                return visualize_nx(path)\n",
    "            else:\n",
    "                return visualize_cys(path)\n",
    "    else:\n",
    "        query = query + \"RETURN p1.id AS identifier1, p1.name as Name1, p2.id AS identifier2, p2.name AS Name2, t.name AS Taxonomy, r.source AS Source, type(r) AS type\"\n",
    "        request = driver.run(query).data()\n",
    "        table = pd.DataFrame(request)\n",
    "        if not table.empty:\n",
    "            table = table.set_index('identifier1').drop_duplicates()\n",
    "            return table\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "from plotly import subplots\n",
    "import plotly.graph_objs as go\n",
    "from plotly.offline import init_notebook_mode, iplot\n",
    "init_notebook_mode(connected=True)\n",
    "\n",
    "def get_simple_scatterplot(data, args):\n",
    "    figure = {}\n",
    "    m = {'size': 15, 'line': {'width': 0}}\n",
    "    text = data.name\n",
    "    if 'colors' in data.columns:\n",
    "        m.update({'color':data['colors'].tolist()})\n",
    "    if 'size' in data.columns:\n",
    "        m.update({'size':data['size'].tolist()})\n",
    "    if 'symbol' in data.columns:\n",
    "        m.update({'symbol':data['symbol'].tolist()})\n",
    "    \n",
    "    annots=[]\n",
    "    if 'annotations' in args:\n",
    "        for index, row in data.iterrows():\n",
    "            name = row['name'].split(' ')[0]\n",
    "            if name in args['annotations']:\n",
    "                annots.append({'x': row['x'], \n",
    "                            'y': row['y'], \n",
    "                            'xref':'x', \n",
    "                            'yref': 'y', \n",
    "                            'text': name, \n",
    "                            'showarrow': False, \n",
    "                            'ax': 55, \n",
    "                            'ay': -1,\n",
    "                            'font': dict(size = 8)})\n",
    "    figure['data'] = [go.Scattergl(x = data.x,\n",
    "                                y = data.y,\n",
    "                                text = text,\n",
    "                                mode = 'markers',\n",
    "                                opacity=0.7,\n",
    "                                marker= m,\n",
    "                                )]\n",
    "                                \n",
    "    figure[\"layout\"] = go.Layout(title = args['title'],\n",
    "                                xaxis= {\"title\": args['x_title']},\n",
    "                                yaxis= {\"title\": args['y_title']},\n",
    "                                margin={'l': 40, 'b': 40, 't': 30, 'r': 10},\n",
    "                                legend={'x': 0, 'y': 1},\n",
    "                                hovermode='closest',\n",
    "                                height=args['height'],\n",
    "                                width=args['width'],\n",
    "                                annotations = annots + [dict(xref='paper', yref='paper', showarrow=False, text='')],\n",
    "                                showlegend=False,\n",
    "                                template='plotly_white'\n",
    "                                )\n",
    "    \n",
    "    return figure\n",
    "\n",
    "def generate_rank(data, args):\n",
    "    num_cols = 2\n",
    "    fig = {}\n",
    "    layouts = []\n",
    "    num_groups = len(data.index.unique())\n",
    "    num_rows = math.ceil(num_groups/num_cols)\n",
    "    fig = subplots.make_subplots(rows=num_rows, cols=num_cols, shared_yaxes=True,print_grid=False)\n",
    "    r = 1\n",
    "    c = 1\n",
    "    range_y = [data['Value'].min(), data['Value'].max()+1]\n",
    "    for index in data.index.unique():\n",
    "        gdata = data.loc[index, :].dropna(how='all').groupby('Name', as_index=False).mean().sort_values(by='Value', ascending=False)\n",
    "        gdata = gdata.reset_index().reset_index()\n",
    "        cols = ['x', 'group', 'name', 'y']\n",
    "        cols.extend(gdata.columns[4:])\n",
    "        gdata.columns = cols\n",
    "        gfig = get_simple_scatterplot(gdata, args)\n",
    "        trace = gfig['data'].pop()\n",
    "        glayout = gfig['layout']['annotations']\n",
    "\n",
    "        for l in glayout:\n",
    "            nlayout = dict(x = l.x,\n",
    "                        y = l.y,\n",
    "                        xref = 'x'+str(c),\n",
    "                        yref = 'y'+str(r),\n",
    "                        text = l.text,\n",
    "                        showarrow = True,\n",
    "                        ax = l.ax,\n",
    "                        ay = l.ay,\n",
    "                        font = l.font,\n",
    "                        align='center',\n",
    "                        arrowhead=1,\n",
    "                        arrowsize=1,\n",
    "                        arrowwidth=1,\n",
    "                        arrowcolor='#636363')\n",
    "            layouts.append(nlayout)\n",
    "        trace.name = index\n",
    "        fig.append_trace(trace, r, c)\n",
    "\n",
    "        if c >= num_cols:\n",
    "            r += 1\n",
    "            c = 1\n",
    "        else:\n",
    "            c += 1\n",
    "    fig['layout'].update(dict(height = args['height'], \n",
    "                            width=args['width'],  \n",
    "                            title=args['title'], \n",
    "                            xaxis= {\"title\": args['x_title'], 'autorange':True}, \n",
    "                            yaxis= {\"title\": args['y_title'], 'range':range_y},\n",
    "                            template='plotly_white'))\n",
    "    fig['layout'].annotations = [dict(xref='paper', yref='paper', showarrow=False, text='')] + layouts \n",
    "\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "@interact\n",
    "def get_species_protein_rank(query_species='9606, 10090', visualize=['Rank', 'Table'], query_proteins='Smad7,Ahnak,Ctsa,MYH1,GOT2'):   \n",
    "    rank = None\n",
    "    args={'x_title':'protein ranking', 'y_title':'intensity', 'title':'Ranking of Proteins', 'height':900, 'width':900}\n",
    "    query= r'MATCH path=(t:Taxonomy)<-[:BELONGS_TO_TAXONOMY]-(pr:Protein)-[r:HAS_QUANTIFIED_PROTEIN]-(p:Project)'\n",
    "    query_species = query_species.split(',')\n",
    "    if '' not in query_species or len(query_species)>1:\n",
    "        query = query + \" WHERE t.id IN [{}] RETURN pr.id AS Identifier, pr.name AS Name, r.value AS Value, t.id AS Taxid\".format(\",\".join(['\"{}\"'.format(s.strip()) for s in query_species]))\n",
    "        request = driver.run(query).data()\n",
    "        table = pd.DataFrame(request)\n",
    "        table = table.set_index('Taxid').drop_duplicates()\n",
    "        if visualize == \"Rank\":\n",
    "            table = table.sort_values('Value',ascending=True)\n",
    "            query_proteins = [p.strip() for p in query_proteins.split(',')]\n",
    "            if '' not in query_proteins or len(query_proteins) > 1:\n",
    "                args['annotations'] = query_proteins\n",
    "            rank = generate_rank(table, args)\n",
    "            iplot(rank)\n",
    "        else:\n",
    "            return table\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "@interact\n",
    "def get_species_protein_rank_with_function(query_species='9606, 10090', visualize=['Rank', 'Table'], query_functions='protein folding'):   \n",
    "    rank = None\n",
    "    args={'x_title':'protein ranking', 'y_title':'intensity', 'title':'Ranking of Proteins', 'height':1200, 'width':900}\n",
    "    query= r'MATCH path=(t:Taxonomy)<-[:BELONGS_TO_TAXONOMY]-(pr:Protein)-[r:HAS_QUANTIFIED_PROTEIN]-(p:Project)'\n",
    "    query_species = query_species.split(',')\n",
    "    if '' not in query_species or len(query_species)>1:\n",
    "        query = query + \" WHERE t.id IN [{}] RETURN pr.id AS Identifier, pr.name AS Name, r.value AS Value, t.id AS Taxid\".format(\",\".join(['\"{}\"'.format(s.strip()) for s in query_species]))\n",
    "        request = driver.run(query).data()\n",
    "        table = pd.DataFrame(request)\n",
    "        query_functions = query_functions.split(';')\n",
    "        if '' not in query_functions or len(query_functions) > 1:\n",
    "            query_functions = r'MATCH path=(t:Taxonomy)<-[:BELONGS_TO_TAXONOMY]-(pr:Protein)-[:ASSOCIATED_WITH]-(f) WHERE t.id IN [{}] AND toLower(f.name) IN [{}] RETURN DISTINCT(pr.name) AS Name, f.name AS Function'.format(\",\".join(['\"{}\"'.format(s.strip()) for s in query_species]),\",\".join(['\"{}\"'.format(f.strip()) for f in query_functions]))\n",
    "            request_functions = driver.run(query_functions).data()\n",
    "            annotations = pd.DataFrame(request_functions)\n",
    "            table = table.set_index('Name').join(annotations.set_index('Name')).reset_index()\n",
    "            args['annotations'] = annotations['Name'].tolist()\n",
    "        table = table.set_index('Taxid').drop_duplicates()\n",
    "        if visualize == \"Rank\":\n",
    "            table = table.sort_values('Value',ascending=True)\n",
    "            rank = generate_rank(table, args)\n",
    "            iplot(rank)\n",
    "        else:\n",
    "            return table\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_barplot(data, args):\n",
    "    figure = {}\n",
    "    figure[\"data\"] = []\n",
    "    if \"group\" in args:\n",
    "        for g in data[args[\"group\"]].unique():\n",
    "            color = None\n",
    "            if 'colors' in args:\n",
    "                if g in args['colors']:\n",
    "                    color = args['colors'][g]\n",
    "            trace = go.Bar(\n",
    "                        x = data.loc[data[args[\"group\"]] == g,args['x']], # assign x as the dataframe column 'x'\n",
    "                        y = data.loc[data[args[\"group\"]] == g, args['y']],\n",
    "                        name = g,\n",
    "                        marker = dict(color=color)\n",
    "                        )\n",
    "            figure[\"data\"].append(trace)\n",
    "    else:\n",
    "        figure[\"data\"].append(\n",
    "                      go.Bar(\n",
    "                            x=data[args['x']], # assign x as the dataframe column 'x'\n",
    "                            y=data[args['y']]\n",
    "                        )\n",
    "                    )\n",
    "    figure[\"layout\"] = go.Layout(\n",
    "                            title = args['title'],\n",
    "                            xaxis={\"title\":args[\"x_title\"]},\n",
    "                            yaxis={\"title\":args[\"y_title\"]},\n",
    "                            height = args['height'],\n",
    "                            width = args['width'],\n",
    "                            annotations = [dict(xref='paper', yref='paper', showarrow=False, text='')],\n",
    "                            template='plotly_white'\n",
    "                        )\n",
    "\n",
    "    return figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@interact\n",
    "def plot_homologs_barplot(query_protein='O14983', query_species='10090,10116'):\n",
    "    query= r'MATCH path=(p1:Protein)-[:IS_ORTHOLOG]-(p2:Protein) WHERE (p1.id=\"{}\" OR p1.name=\"{}\") WITH collect(p1.id) AS ids1, collect(p2.id) AS ids2 MATCH (t:Taxonomy)-[:BELONGS_TO_TAXONOMY]-(pr:Protein)-[r:HAS_QUANTIFIED_PROTEIN]-(p:Project) WHERE pr.id IN ids1 OR pr.id IN ids2'.format(query_protein, query_protein)\n",
    "    query_species = query_species.split(',')\n",
    "    if '' not in query_species:\n",
    "        query = query + \" AND t.id IN [{}]\".format(\",\".join(['\"{}\"'.format(s.strip()) for s in query_species]))\n",
    "    query = query + \" RETURN pr.id AS Identifier, r.value AS Value, t.name AS Taxid\"\n",
    "    request = driver.run(query).data()\n",
    "    table = pd.DataFrame(request)\n",
    "    title= \"Plot Homologs {}\".format(query_protein)\n",
    "    figure = get_barplot(table, args={'group':'Taxid', 'x':'Identifier', 'y':'Value',\n",
    "                                      'x_title':'Homologs',\n",
    "                                      'y_title':'Intensity',\n",
    "                                      'height':600,\n",
    "                                      'width':1000,\n",
    "                                      'title':title})\n",
    "    iplot(figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b0735b84604f48228b9c459d40ba4fd8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Text(value='O14983', description='query_protein'), FloatSlider(value=0.1, description='s…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import ast          \n",
    "def give_cytonet_style(node_colors):\n",
    "    color_selector = \"{'selector': '[name = \\\"KEY\\\"]', 'style': {'background-color': 'VALUE'}}\"\n",
    "    stylesheet=[{'selector': 'node', 'style': {'label': 'data(name)','width':5, 'height':5,'font-size':1,}}, \n",
    "                {'selector':'edge','style':{'curve-style': 'bezier', 'width':0.5}}]\n",
    "    \n",
    "    stylesheet.extend([{'selector':'[width < 1.1]', 'style':{'line-color':'#ef3b2c', 'line-style': 'dashed', 'width':'width'}},{'selector':'[width = 1.1]', 'style':{'line-color':'#2171b5','width':'width'}},{'selector':'[width > 1.1]', 'style':{'line-color':'#ae017e','width':'width'}}])\n",
    "\n",
    "    for k,v in node_colors.items():\n",
    "        stylesheet.append(ast.literal_eval(color_selector.replace(\"KEY\", k.replace(\"'\",\"\\'\")).replace(\"VALUE\",v)))\n",
    "    \n",
    "    return stylesheet\n",
    "\n",
    "@interact\n",
    "def predict_protein_function(query_protein='O14983', score_cutoff=0.1, visualize=['Cytoscape', 'Table', 'Networkx']):   \n",
    "    net = None\n",
    "    query_total_homologs = r'MATCH path=(p1:Protein)-[:IS_ORTHOLOG]-(p2:Protein)-[:HAS_QUANTIFIED_PROTEIN]-() WHERE (p1.id=\"{}\" OR p1.name=\"{}\") AND p2.id<>p1.id RETURN count(DISTINCT(p2.id)) AS total_homologs'.format(query_protein, query_protein)\n",
    "    query= r'MATCH path=(p1:Protein)-[:IS_ORTHOLOG]-(p2:Protein)-[:HAS_QUANTIFIED_PROTEIN]-() WHERE (p1.id=\"{}\" OR p1.name=\"{}\") AND p2.id<>p1.id WITH p1, p2, collect(p2.id) AS ids2 MATCH (p1)-[:ASSOCIATED_WITH]-(f:Biological_process) WITH collect(f.id) AS functions, p1,p2, ids2 MATCH (p2)-[:ASSOCIATED_WITH]-(f:Biological_process) WHERE NOT f.id IN functions RETURN p1.id AS Identifier, p1.name AS Name, f.id AS Function_id, f.name AS Function, count(f.id) AS Num_Orthologs, collect(p2.id) AS Orthologs ORDER BY Num_Orthologs DESC'.format(query_protein, query_protein)\n",
    "    query_functions= r'MATCH (p1)-[:ASSOCIATED_WITH]-(f:Biological_process) WHERE (p1.id=\"{}\" OR p1.name=\"{}\") RETURN p1.id AS Identifier, p1.name AS Name, f.id AS Function_id, f.name AS Function'.format(query_protein, query_protein)\n",
    "    \n",
    "    total_homologs = driver.run(query_total_homologs).data()[0]['total_homologs']\n",
    "    request = driver.run(query).data()\n",
    "    table = pd.DataFrame(request)\n",
    "    if not table.empty:\n",
    "        table['total_orthologs'] = total_homologs\n",
    "        table['score'] = table['Num_Orthologs']/total_homologs\n",
    "        table = table[table['score'] >= score_cutoff]\n",
    "        if visualize == \"Table\":\n",
    "            return table\n",
    "        elif visualize == \"Networkx\" or visualize == \"Cytoscape\":\n",
    "            nodes = set()\n",
    "            rels = set()\n",
    "            functions = set()\n",
    "            node_colors = {}\n",
    "            for index, row in table.iterrows():\n",
    "                source = row['Function'].title()\n",
    "                functions.add(row['Function_id'])\n",
    "                target = row['Name']\n",
    "                node_colors[source] = '#ef3b2c'\n",
    "                relationship = ('predicted', row['score'], '#ef3b2c')\n",
    "                nodes.update([source, target])\n",
    "                rels.add((source, target, relationship))\n",
    "            request_annotation = driver.run(query_functions).data()\n",
    "            annotations = pd.DataFrame(request_annotation)\n",
    "            if not annotations.empty:\n",
    "                for index, row in annotations.iterrows():\n",
    "                    source = row['Function'].title()\n",
    "                    functions.add(row['Function_id'])\n",
    "                    target = row['Name']\n",
    "                    node_colors[source] = '#2171b5'\n",
    "                    relationship = ('annotated', 1.1, '#2171b5')\n",
    "                    nodes.update([source, target])\n",
    "                    rels.add((source, target, relationship))\n",
    "            functions = \",\".join(['\"{}\"'.format(f) for f in functions])\n",
    "            query_functions_rels = r'MATCH (f1:Biological_process)-[:HAS_PARENT]-(parent:Biological_process)-[:HAS_PARENT]-(f2:Biological_process) WHERE f1.id IN [{}] AND f2.id IN [{}] AND f1.id<>f2.id RETURN f1.name AS source1, parent.name AS parent, f2.name AS source2'.format(functions, functions)\n",
    "            request_hierarchy = driver.run(query_functions_rels).data()\n",
    "            hierarchy = pd.DataFrame(request_hierarchy)\n",
    "            if not hierarchy.empty:\n",
    "                for index, row in hierarchy.iterrows():\n",
    "                    source1 = row['source1'].title()\n",
    "                    target = row['parent']\n",
    "                    source2 = row['source2'].title()\n",
    "                    node_colors[target] = '#ae017e'\n",
    "                    relationship = ('hierarchy', 2, '#ae017e')\n",
    "                    nodes.update([target])\n",
    "                    rels.add((source1, target, relationship))\n",
    "                    rels.add((source2, target, relationship))\n",
    "\n",
    "            if len(nodes)> 0:\n",
    "                G = nx.Graph()        \n",
    "                G.add_nodes_from(nodes, color='blue')\n",
    "                for s,t,attrs in rels:\n",
    "                    G.add_edge(s,t,label=attrs[0], width=attrs[1], color=attrs[2])\n",
    "                if visualize == \"Networkx\":\n",
    "                    return visualize_nx(G)\n",
    "                else:\n",
    "                    cysG, layout, _ = networkx_to_cytoscape(G)\n",
    "                    stylesheet = give_cytonet_style(node_colors)\n",
    "                    cyvis = Cytoscape(data={'elements':cysG}, visual_style=stylesheet, layout=layout)\n",
    "                    return cyvis\n",
    "            \n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ebbb9f232e67467d8fc33c89171aa6ca",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Text(value='9606, 10090, 10116, 224325, 572546, 589924, 272569, 1435377, 273057, 9031, 3…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import plotly.express as px\n",
    "@interact\n",
    "def get_functional_groups_overview(query_species='9606, 10090, 10116, 224325, 572546, 589924, 272569, 1435377, 273057, 9031, 3847, 1620505, 55529', \n",
    "                                   query_functions='carbohydrate metabolic process, glycolytic process, ion transport, oxidation-reduction process,photosynthesis,protein folding,proteolysis,translation,translational elongation',\n",
    "                                   visualize=['polar', 'table']):\n",
    "    query = r'MATCH (t:Taxonomy)-[:BELONGS_TO_TAXONOMY]-(p:Protein)-[r:HAS_QUANTIFIED_PROTEIN]-() WHERE t.id IN [{}] WITH t, p, r MATCH (t)-[]-(p)-[:ASSOCIATED_WITH]-(bp:Biological_process) WHERE toLower(bp.name) IN [{}] RETURN t.name AS Taxid, r.value AS Intensity, bp.name AS Function'\n",
    "    query_species = query_species.split(',')\n",
    "    query_functions = query_functions.split(',')\n",
    "    if '' not in query_species or len(query_species)>1:\n",
    "        query_species = \",\".join(['\"{}\"'.format(s.strip()) for s in query_species])\n",
    "        if '' not in query_functions or len(query_functions)>1:\n",
    "            query_functions = \",\".join(['\"{}\"'.format(f.strip()) for f in query_functions])\n",
    "            query = query.format(query_species, query_functions)\n",
    "    \n",
    "    request = driver.run(query).data()\n",
    "    table = pd.DataFrame(request)\n",
    "    if not table.empty:\n",
    "        if visualize == \"polar\":\n",
    "            df = table.groupby(['Taxid', 'Function'])\n",
    "            list_cols = []\n",
    "            for group in df.groups:\n",
    "                mean = df.get_group(group).mean()[\"Intensity\"]\n",
    "                taxid, function = group\n",
    "                list_cols.append([taxid, function, mean])\n",
    "            summ_df = pd.DataFrame(list_cols, columns=['Taxid', 'Function', 'avg Intensity'])\n",
    "            summ_df['Function'] = summ_df['Function'].apply(lambda x: x.title())\n",
    "            fig = px.line_polar(summ_df, r=\"avg Intensity\", theta=\"Taxid\", color=\"Function\", line_close=True,\n",
    "                    title='Mean Expression Levels of Functional Groups',\n",
    "                    color_discrete_sequence=px.colors.sequential.Plasma[-2::-1], width=1000)\n",
    "            iplot(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
